pkg/storage/cacher.go

```diff
@@ -859,10 +859,19 @@ func (c *cacheWatcher) sendWatchCacheEvent(event *watchCacheEvent) {
 	case !curObjPasses && oldObjPasses:
 		watchEvent = watch.Event{Type: watch.Deleted, Object: object}
 	}
+
+	// We don't want to block on c.result if c.done is already closed.
+	//
+	// TODO: This doesn't protect with the situation when c.result is
+	// already full and we are handing on it and in the meantime c.done
+	// is being closed and noone will clear the result channel.
+	// However, we can't heve select of two cases, as this may lead to
+	// delivering random part of events after c.done is closed but we
+	// are still sending them for a moment.
 	select {
-	case c.result <- watchEvent:
-	// don't block on c.result if c.done is closed
 	case <-c.done:
+	default:
```

Comment:
I'm happy to change the comment, but in my opinion it's not misleading. So if you can suggest something else, I'm happy to change that.

Regarding your proposal about changing to use lock - I don't see how this is fixing the problem. The watcher can still be stopped right after you are calling c.Unlock, which will again decrease the probability but will not fix the problem completely. And you are adding Lock for every (watch event , watcher) pair.